File: apps/deep-agent-service/src/deep_agent_service/standard_sql.py
"""Authorized SQL Toolkit composition. No model input can choose a database."""
import asyncio
import copy
import json
import os
from pathlib import Path
from urllib.parse import quote,urlsplit
import httpx
from jsonschema import Draft7Validator,FormatChecker
from langchain.tools import ToolRuntime
from langchain_core.tools import StructuredTool
from langchain_community.agent_toolkits.sql.toolkit import SQLDatabaseToolkit
from .standard_memory import _identity
from .standard_sql_database import DialectOnlyDatabase,StandardSqlError,call_sql_database

_DIR=Path(__file__).parent/'generated'
_SCHEMA=json.loads((_DIR/'standard_sql_schema.json').read_text())
_NATIVE=json.loads((_DIR/'standard_sql_tools.json').read_text())
_V={key:Draft7Validator(_SCHEMA[key],format_checker=FormatChecker()) for key in ('input','output','sources')}

async def _source(name,args,runtime):
    scope,callback=_identity(runtime)
    base=callback['base_url'].rstrip('/');url=urlsplit(base)
    if url.scheme not in ('http','https') or not url.hostname or url.username or url.password or url.query or url.fragment or not callback['key']:raise ValueError()
    body={'orgId':scope['orgId'],'userId':scope['userId'],'attemptId':callback['attempt_id'],'leaseEpoch':callback['lease_epoch'],
          'toolCallId':runtime.tool_call_id,'toolName':name,'toolArgs':args}
    if callback.get('permission_request_id'):body['permissionRequestId']=callback['permission_request_id']
    _V['input'].validate(body)
    async with asyncio.timeout(5),httpx.AsyncClient(timeout=5,follow_redirects=False,trust_env=False) as client:
        async with client.stream('POST',f"{base}/internal/agent-runs/{quote(callback['run_id'],safe='')}/sql/source/check",json=body,
                headers={'x-deep-agent-internal-key':callback['key'],'accept-encoding':'identity'}) as response:
            if response.status_code!=200 or response.headers.get('content-encoding','identity')!='identity':raise ValueError()
            content=bytearray()
            async for chunk in response.aiter_raw():
                if len(content)+len(chunk)>4096:raise ValueError()
                content.extend(chunk)
            result=json.loads(content);_V['output'].validate(result)
    sources=json.loads(os.environ.get('STANDARD_SQL_SOURCES','{}'));_V['sources'].validate(sources)
    source=sources[result['dataSourceId']]
    dsn=urlsplit(source['dsn'])
    if dsn.scheme!='postgresql+psycopg' or not dsn.hostname or not dsn.username or not dsn.username.startswith('wsx_sql_ro_') or dsn.query or dsn.fragment:
        raise ValueError()
    if dsn.path.lstrip('/') in source['applicationDatabases']:raise ValueError()
    return source

def standard_sql_tools(model):
    native={tool.name:tool for tool in SQLDatabaseToolkit(db=DialectOnlyDatabase(),llm=model).get_tools()}
    def build(name):
        schema=copy.deepcopy(_NATIVE['tools'][name]);schema['additionalProperties']=False
        for field in schema['properties'].values():field['maxLength']=_SCHEMA['limits']['maxArgumentChars']
        async def invoke(runtime:ToolRuntime,**kwargs):
            try:
                Draft7Validator(schema).validate(kwargs)
                source=await _source(name,kwargs,runtime)
                if name=='sql_db_query_checker':
                    # Real upstream async model checker. No engine/reflection/query.
                    async with asyncio.timeout(_SCHEMA['limits']['databaseTimeoutMs']/1000):output=await native[name]._arun(**kwargs)
                    if len(output.encode())>_SCHEMA['limits']['maxOutputBytes']:raise ValueError()
                    return output
                return await call_sql_database(source,name,kwargs,model)
            except StandardSqlError:raise
            except Exception:raise StandardSqlError('sql_unavailable_or_refused') from None
        def sync(runtime:ToolRuntime,**kwargs):return asyncio.run(invoke(runtime,**kwargs))
        return StructuredTool(name=name,description=native[name].description,args_schema=schema,func=sync,coroutine=invoke)
    return [build(name) for name in native]

File: packages/contracts/src/standard-sql.ts
import {z} from 'zod';
import native from './generated/standard-sql-tools.json';
export const SQL_LIMITS={maxArgumentChars:16384,maxRows:100,maxRowChars:4096,maxOutputBytes:65536,statementTimeoutMs:2000,databaseTimeoutMs:8000,maxConcurrentOperations:8} as const;
const names=Object.keys(native.tools) as [string,...string[]];
export const StandardSqlToolName=z.enum(names);
const id=z.string().min(1).max(256);
const identifier=z.string().regex(/^[a-zA-Z_][a-zA-Z0-9_]{0,62}$/);
export const StandardSqlSourceCheckInput=z.object({orgId:id,userId:id,attemptId:id,leaseEpoch:z.number().int().positive(),
 toolCallId:id,permissionRequestId:z.string().uuid().optional(),toolName:StandardSqlToolName,toolArgs:z.record(z.string().max(SQL_LIMITS.maxArgumentChars))}).strict();
export const StandardSqlSourceCheckOutput=z.object({dataSourceId:identifier}).strict();
export const StandardSqlBindings=z.array(z.object({orgId:id,userIds:z.array(id).min(1).max(256),dataSourceId:identifier}).strict()).max(256);
export const StandardSqlSources=z.record(identifier,z.object({dsn:z.string().min(1).max(4096),sslMode:z.enum(['verify-full','verify-ca','require']),schema:identifier,views:z.array(identifier).min(1).max(64),
 applicationDatabases:z.array(z.string().min(1).max(128)).min(1).max(16)}).strict());
/** Native upstream field names/requiredness; transport adds a length ceiling and rejects unknown keys. */
export function validateSqlToolArgs(toolName:string,args:unknown):Record<string,string>{
 const schema=(native.tools as Record<string,{properties:Record<string,unknown>;required?:string[]}>)[toolName];
 if(!schema)throw new Error('unknown SQL tool');
 return z.object(Object.fromEntries(Object.keys(schema.properties).map(key=>[key,schema.required?.includes(key)?z.string().max(SQL_LIMITS.maxArgumentChars):z.string().max(SQL_LIMITS.maxArgumentChars).optional()]))).strict().parse(args) as Record<string,string>;
}
